core/src/main/java/org/opensearch/sql/planner/physical/TakeOrderedOperator.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.planner.physical;
+
+import static org.opensearch.sql.ast.tree.Sort.NullOrder.NULL_FIRST;
+import static org.opensearch.sql.ast.tree.Sort.SortOrder.ASC;
+
+import com.google.common.collect.Ordering;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Singular;
+import lombok.ToString;
+import org.apache.commons.lang3.tuple.Pair;
+import org.opensearch.sql.ast.tree.Sort.SortOption;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.utils.ExprValueOrdering;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.planner.physical.TakeOrderedOperator.ExprOrdering.ExprOrderingBuilder;
+
+/**
+ * TakeOrdered Operator. This operator will sort input data as the order of {@link this#sortList}
+ * specifies and return {@link this#limit} rows from the {@link this#offset} index.
+ *
+ * <p>Functionally, this operator is a combination of {@link SortOperator} and {@link
+ * LimitOperator}. But it can reduce the time complexity from O(nlogn) to O(n), and memory from O(n)
+ * to O(k) due to use guava {@link com.google.common.collect.Ordering}.
+ *
+ * <p>Overall, it's an optimization to replace `Limit(Sort)` in physical plan level since it's all
+ * about execution. Because most execution engine may not support this operator, it doesn't have a
+ * related logical operator.
+ */
+@ToString
+@EqualsAndHashCode(callSuper = false)
+public class TakeOrderedOperator extends PhysicalPlan {
+  @Getter private final PhysicalPlan input;
+
+  @Getter private final List<Pair<SortOption, Expression>> sortList;
+  @Getter private final Integer limit;
+  @Getter private final Integer offset;
+  @EqualsAndHashCode.Exclude private final ExprOrdering ordering;
+  @EqualsAndHashCode.Exclude private Iterator<ExprValue> iterator;
+
+  /**
+   * TakeOrdered Operator Constructor.
+   *
+   * @param input input {@link PhysicalPlan}
+   * @param limit the limit value from LimitOperator
+   * @param offset the offset value from LimitOperator
+   * @param sortList list of sort field from SortOperator
+   */
+  public TakeOrderedOperator(
```

Comment:
is it possible to combine SortOperator and TakeOrderedOperator. Both operator has next interface return the sorted valued, the difference is TakeOrderedOperator is more efficiency with limit + offset.

Reply:
TakeOrderedOperator could cover the usage of SortOperator functionally if we set limit to MAX_VALUE and offset to 0.

But it may be a little bit confusing to do that because TakeOrderedOperator aims to do calculating for taking Top-K. So these 2 operators should have different operator name and parameters when users look into plan explain. 

They may be confusing to see explain with TakeOrderedOperator whose parameters contains limit =  MAX_VALUE and offset = 0, if they just want to sort without limit. In turn, it's also confusing to see explain with only SortOperator  whose parameters contain limit and offset while not seeing LimitOperator, if they want to sort + limit k.

Reply:
i see, make sense